tools/clang/unittests/HLSLExec/LongVectors.h

```diff
@@ -199,62 +206,189 @@ void FillLongVectorDataFromShaderBuffer(MappedData &ShaderBuffer,
+                   OpTypeString.c_str());
+
+  if (std::is_same_v<T, LongVectorBinaryOpType>)
+    return static_cast<T>(LongVectorBinaryOpType_EnumValueCount);
+  else if (std::is_same_v<T, LongVectorUnaryOpType>)
+    return static_cast<T>(LongVectorUnaryOpType_EnumValueCount);
+}
+
+static const LongVectorOpTypeStringToEnumValue
+    LongVectorBinaryOpTypeStringToEnumMap[] = {
+        {L"LongVectorBinaryOpType_ScalarAdd", LongVectorBinaryOpType_ScalarAdd},
+        {L"LongVectorBinaryOpType_ScalarMultiply",
+         LongVectorBinaryOpType_ScalarMultiply},
+        {L"LongVectorBinaryOpType_Multiply", LongVectorBinaryOpType_Multiply},
+        {L"LongVectorBinaryOpType_Add", LongVectorBinaryOpType_Add},
+        {L"LongVectorBinaryOpType_Min", LongVectorBinaryOpType_Min},
+        {L"LongVectorBinaryOpType_Max", LongVectorBinaryOpType_Max},
+        {L"LongVectorBinaryOpType_ScalarMin", LongVectorBinaryOpType_ScalarMin},
+        {L"LongVectorBinaryOpType_ScalarMax", LongVectorBinaryOpType_ScalarMax},
+};
+
+static_assert(_countof(LongVectorBinaryOpTypeStringToEnumMap) ==
+                  LongVectorBinaryOpType_EnumValueCount,
+              "LongVectorBinaryOpTypeStringToEnumMap size mismatch. Did you "
+              "add a new enum value?");
+
+LongVectorBinaryOpType
+GetLongVectorBinaryOpType(const std::wstring &OpTypeString) {
+  return GetLongVectorOpType<LongVectorBinaryOpType>(
+      LongVectorBinaryOpTypeStringToEnumMap, OpTypeString,
+      std::size(LongVectorBinaryOpTypeStringToEnumMap));
+}
+
+enum LongVectorUnaryOpType {
+  LongVectorUnaryOpType_Clamp,
+  LongVectorUnaryOpType_Initialize,
+  LongVectorUnaryOpType_EnumValueCount
 };
 
+static const LongVectorOpTypeStringToEnumValue
+    LongVectorUnaryOpTypeStringToEnumMap[] = {
+        {L"LongVectorUnaryOpType_Clamp", LongVectorUnaryOpType_Clamp},
+        {L"LongVectorUnaryOpType_Initialize", LongVectorUnaryOpType_Initialize},
+};
+
+static_assert(_countof(LongVectorUnaryOpTypeStringToEnumMap) ==
+                  LongVectorUnaryOpType_EnumValueCount,
+              "LongVectorUnaryOpTypeStringToEnumMap size mismatch. Did you add "
+              "a new enum value?");
+
+LongVectorUnaryOpType
+GetLongVectorUnaryOpType(const std::wstring &OpTypeString) {
+  return GetLongVectorOpType<LongVectorUnaryOpType>(
+      LongVectorUnaryOpTypeStringToEnumMap, OpTypeString,
+      std::size(LongVectorUnaryOpTypeStringToEnumMap));
+}
+
+template <typename T> std::vector<T> GetInputValueSetByKey(std::wstring Key) {
+
+  if constexpr (std::is_same_v<T, HLSLBool_t>)
+    return std::vector<T>(DefaultInputValueSet_bool.at(Key).begin(),
+                          DefaultInputValueSet_bool.at(Key).end());
+  else if constexpr (std::is_same_v<T, HLSLHalf_t>)
+    return std::vector<T>(DefaultInputValueSet_float16.at(Key).begin(),
+                          DefaultInputValueSet_float16.at(Key).end());
+  else if constexpr (std::is_same_v<T, float>)
```

Comment:
This pattern of using `if constexpr (std::is_same_v<T, float>)` I think might be leading you into some directions that result in more complicated / verbose code than necessary.

Here's an approach that uses more traditional template specialization: https://godbolt.org/z/15qjobn9j